tools/generator/dfg/sam/avr_device_tree.py

```diff
@@ -0,0 +1,261 @@
+            # module SPI has two instances USART0_SPI, USART1_SPI, wtf?
+            if tmp['module'] == 'spi' and tmp['instance'].startswith("usart"): continue;
+            if tmp['module'] == 'port':
+                ports.append(tmp)
+            else:
+                modules.append(tmp)
+        p['modules'] = sorted(list(set([(m['module'], m['instance']) for m in modules])))
+
+        raw_signals = device_file.query("//peripherals/module/instance/signals/signal")
+        modules_only = not len(raw_signals)
+        p['modules_only'] = modules_only
+
+        if modules_only:
+            # Parse GPIOs manually from mask
+            for port in ports:
+                port = port['instance'][-1:]
+                port_mask = device_file.query("//modules/module[@name='PORT']/register-group[@name='PORT{0}']/register[@name='PORT{0}']/@mask".format(port.upper()))
+                if len(port_mask) == 0:
+                    # The port mask is split up
+                    port_mask = device_file.query("//modules/module[@name='PORT']/register-group[@name='PORT{0}']/register[@name='PORT{0}']/bitfield/@mask".format(port.upper()))
+                    port_mask = sum([int(mask, 16) for mask in port_mask])
+                else:
+                    port_mask = int(port_mask[0], 16)
+                for pos in range(8):
+                    if (1 << pos) & port_mask:
+                        gpios.append((port, str(pos)))
+            gpios = sorted(gpios)
+        else:
+            for s in raw_signals:
+                tmp = {'module': s.getparent().getparent().getparent().get('name').lower(),
+                       'instance': s.getparent().getparent().get('name').lower()}
+                tmp.update({k:v.lower() for k,v in s.items()})
+                if tmp['group'] in ['p', 'pin'] or tmp['group'].startswith('port'):
+                    gpios.append(tmp)
+                else:
+                    signals.append(tmp)
+            gpios = sorted([(g['pad'][1], g['pad'][2]) for g in gpios])
+
+        # Signal information is missing
+        p['signals'] = signals
+        p['gpios'] = gpios
+
+        LOGGER.debug("Found GPIOs: [%s]", ", ".join([p.upper() + i for p,i in p['gpios']]))
+        LOGGER.debug("Available Modules are:\n" + AVRDeviceTree._modulesToString(p['modules']))
+
+        interrupts = []
+        for i in device_file.query("//interrupts/interrupt"):
+            interrupts.append({'position': i.get('index'), 'name': i.get('name')})
+        p['interrupts'] = interrupts
+
+        return p
+
+    @staticmethod
+    def _modulesToString(modules):
+        string = ""
+        mods = sorted(modules)
+        char = mods[0][0][0:1]
+        for module, instance in mods:
+            if not instance.startswith(char):
+                string += "\n"
+            string += instance + " \t"
+            char = instance[0][0:1]
+        return string
+
+    @staticmethod
+    def _device_tree_from_properties(p):
```

Comment:
This is the function that takes this vendor specific internal data and formats it into our device file format, which is tree based.
This tree has a few intrinsic properties:
- it is **sortable** so that the output is reproducible (very important for checking if something changed via git),
- it is **mergeable**, so that we can reduce the data duplication and overlay multiple data trees of very similiar devices.

This allows us to reduce the number of unique data tree by ~20 fold, which makes it significantly easier to perform manual spot checks to verify the data pipeline.
See also: https://blog.salkinium.com/modm-devices/#encoding-commonality for explanation and examples.